content/post/20220601-quantum-compiler.md

```diff
@@ -0,0 +1,352 @@
+H 0       H 0       S 0       H 0       S 1       T 1       T 1       S 1       
+T 0       T 0       H 0       T 0       H 1       S 1       S 1       S 1       
+S 0       S 0       T 0       S 0       T 1       H 1       H 1       S 1       
+H 0       H 0       S 0       H 0       S 1       T 1       T 1       CNOT 0 1  
+T 0       T 0       H 0       T 0       H 1       S 1       H 1       H 0       
+S 0       S 0       T 0       H 0       T 1       H 1       T 1       T 0       
+H 0       H 0       H 0       S 0       H 1       T 1       H 1       T 1       
+T 0       T 0       T 0       S 0       T 1       S 1       T 1       S 1       
+H 0       S 0       S 0       S 0       H 1       H 1       H 1       S 1       
+T 0       H 0       H 0       H 0       T 1       T 1       T 1       S 1       
+S 0       T 0       T 0       S 0       S 1       S 1       S 1                 
+H 0       S 0       S 0       S 0       H 1       H 1       H 1                 
+T 0       H 0       H 0       S 0       T 1       T 1       T 1                 
+S 0       T 0       T 0       S 1       H 1       S 1       S 1                 
+H 0       S 0       H 0       H 1       T 1       H 1       H 1                 
+T 0       H 0       T 0       S 1       S 1       T 1       T 1              
+```
+
+Notice how every operation in the output is either `H`, `S`, `T`, or `CNOT`. But also notice how there are a _lot_ of operations—290 of them to be exact. There are only two differences in how we run the compiler between this and the previous IBM-chip compilation:
+
+1. We had to specify the chip `discrete2`. (Quilc will be happy to construct a chip of any requested shape, size, and native operations; `discrete2` is just a convenient built-in for testing.)
+2. We had to specify a tolerance, here a discrete compilation accuracy within 0.1% for each approximation made.
+
+If we request more precision, we notice an increase in operation count.
+
+```
+$ echo 'PRAGMA TOLERANCE "0.0000001d0"; XY(pi/3) 1 0' | ./quilc --isa discrete2 | wc -l
+     540
+```
+
+Despite being 1.8x the number of operations more, we've also increased precision by 10,000x. This is in line with the efficiency of the Selinger algorithm: an exponential increase in precision does _not_ mean an exponential increase in compiled output. Here, we can see the trend in compilation efficiency for different precisions
+
+***
+
+| Precision | Operation Count |
+| --------- | --------------- |
+| 0.1       | 152             |
+| 0.01      | 232             |
+| 0.001     | 290             |
+| 0.0001    | 354             |
+| 0.00001   | 404             |
+| 0.000001  | 502             |
+| 0.0000001 | 540             |
+
+***
+
+One particularly nice thing going on is that the whole machinery quilc uses top optimize programs automatically gets applied. It's not _just_ a discrete compiler, it's an optimizing compiler with support for discrete operation sets.
+
+## Inaccuracy and validation
+
```

Comment:
> Here is some stuff I wrote for this section. Feel free to add, remove, or edit whatever you see fit.

With this implementation the only operator being approximated is $\mathrm{RZ}_\theta$, which means the error will increase whenever two $\mathrm{RZ}$ approximations are used in the same computation. Our `TOLERANCE` pragma is asserting the precision of individual approximations, not whole quilc programs. While there are rules that give insight into how these errors accumulate, such notions are not yet readily available to plug into quilc. To get around this requires some trial and error from the user by attempting compilations with increasing precision until a tolerable error is achieved. To help with this, passing `-Pm` to `quilc` will print out the whole program's error. Nonetheless, we are guaranteed the precision of our approximation ($U$) is directly related to a tolerance of $\epsilon$ by:

$$ \lvert \lvert(\mathrm{RZ}_\theta - U) \mathrm{v} \rvert \rvert \leq \epsilon \lvert \lvert \mathrm{v} \rvert \rvert$$
where $0<\epsilon < 0.5$,  $\mathrm{v}$ is a two dimensional complex vector, and $\lvert \lvert \cdot \rvert \rvert$ is the vector magnitude. We can read the above expression as "the distance between $\mathrm{RZ}_\theta$ and $U$ is less than or equal to $\epsilon$."

Quantum states are probabilistic, and quantum operators affect a state's probabilities. For a qubit, these are the probabilities of whether we measure 0 or 1. If two quantum operators are *close*, then the probabilities of the qubits they act on should be similiar. To see this in action we can approximate a 3-qubit W state with various tolerance values and measure it a few thousand times on a simulated quantum computer (i.e. [QVM](https://github.com/quil-lang/qvm)). Just how a coin toss has a 50–50 distribution, each qubit in a W state has a 1/3 chance of being 1 when measured. Expectedly, as we increase precision, the closer we get to an even distribution.

***
| Tolerance | 001 |  010 | 100 | $\chi^2$ |
|------|--------|--------|--------|---------|
|$2^{-1}$  | 11.35% | 44.35% | 44.30% | 2.2E+05 |
|$2^{-2}$  | 34.89% | 32.55% | 32.57% | 1.1E+03 |
|$2^{-4}$  | 33.50% | 33.29% | 33.21% | 1.4E+01 | 
|$2^{-8}$  | 33.21% | 33.42% | 33.37% | 7.1E+00 |
|$2^{-16}$ | 33.43% | 33.28% | 33.28% | 4.4E+00 |
|$2^{-32}$ | 33.40% | 33.31% | 33.29% | 1.9E+00 | 
|$2^{-64}$ | 33.36% | 33.31% | 33.33% | 2.8E-01 |
***

Note that chi-squared ($\chi^2$) quantifies how far off the the results are from the expected distribution.